### ISSUE TITLE
Debate: Decoupling response and serialization

### ISSUE THREAD

**Issue by vladmunteanu** [vladmunteanu/starlette-jsonapi#8]
There are a number of things we could implement much easier if serialization and response would be decoupled, such that `BaseResource.serialize` returns a `dict`, not a `JSONAPIResponse`.

1) We add a `handle_response(data: Any, *args, **kwargs) -> JSONAPIResponse` method on `BaseResource`, which gets called from `BaseResource.handle_request` with the result of the handler (`BaseResource.get`, `BaseResource.post` etc) and returns a `JSONAPIResponse` object for `data`. 
It also means we're changing the signature of handlers to `Union[dict, JSONAPIResponse]`.

2) We add a `to_response(data: Any, *args, **kwargs) -> JSONAPIResponse` method that gets called explicitly from the handler and wraps `data` in a `JSONAPIResponse`.
The signature of `BaseResource.serialize` would change to `serialize(.....) -> dict`.

Both methods would be accessible to subclasses to override and add custom implementation if needed. Main difference is the implicit behavior of 1).
Also, we would be able to separate arguments needed for serialization and arguments needed to generate responses, in the end possibly allowing more access to the underlying libraries via `args` and `kwargs`.

I personally like 2), but resources would probably need to do something like `await self.to_response(self.serialize(data))`.

**Comment by vladmunteanu** [comment]
If option 2) is the way to go, I propose another helper `BaseResource.serialize_to_response` which does the serialization and returns the response (pretty much what `BaseResource.serialize` does now).

**Comment by khornberg** [comment]
I think I agree with option 2.

However, instead of adding another method of `to_repsonse`, the response in the view can be

```
        async def post(self, *args, **kwargs) -> Response:
            return await JSONAPIResponse(content=self.serialize(...))
```

That separates the concern of serializing the data and forming a response.
The user can either subclass `JSONAPIResponse` or return something else entirely.

It also simplifies the changes internally as well

**Comment by vladmunteanu** [comment]
@khornberg Thank you very much for your input! 
Although I tend to agree with you, and started the #9 PR to experiment with this, I would still like to have `to_response`, in case subclasses would like to override it and do custom things, such as adding more headers maybe. 
People will still be able to do what you described, and I will too probably if I would ever need to subclass `JSONAPIResponse`.

My main concern when it comes to simplicity for this stuff, is that `serialize` will likely get more complicated as we continue to improve support for jsonapi features (pagination, meta and jsonapi objects for example). There's a chance that response related logic will also grow.

That being said, I'm not 100% happy with #9 because of the kwargs handling in `serialize_to_response`, so I will definitely revisit that soon and might get rid of `serialize_to_response` entirely for the sake of simplicity.

**Comment by pOctav** [comment]
I agree with this point as well. I think this method offers a good balance between customizability and ease of use. Like @vladmunteanu said, the option to directly return a `JSONAPIResponse` in the resource still remains available, whereas `to_response` offers both a quick way to wrap any dict as a response, and it also allows for overriding in subclasses

**Comment by vladmunteanu** [comment]
I've updated the #9 PR and removed the `serialize_to_response` method. 
Going to look into examples and documentation next.

**Comment by vladmunteanu** [comment]
Fixed in 1.0.0
